### PR TITLE
[2.8] Fix race in start hook after reboot test

### PR DESCRIPTION
**Pull request by achilleasa**
This PR backports the fixes from #12250 and adds extra logic to the hook tests to wait for the systemd files to be written to disk prior to attempting to restart the unit's agent.

## QA steps

```sh
cd tests
./main.sh -v hooks
```

### CODE REVIEW

File: tests/includes/wait-for.sh [hunk 1]
```diff
@@ -158,6 +158,38 @@ wait_for_machine_netif_count() {
...
+# wait_for_subordinate_count blocks until the number of subordinates
+# to the desired unit becomes equal to the desired value.
+#
+# ```
+# wait_for_subordinate_count <applicaiton name> <principal unit num> <count>
+#
+# example:
+# wait_for_subordinate_count mysql 0 3
+# ```
+wait_for_subordinate_count() {
```
Comment (achilleasa) [inline tests/includes/wait-for.sh:170]: This has been back-ported from #12250

File: tests/suites/hooks/reboot.sh [hunk 2]
```diff
@@ -26,6 +26,7 @@ run_start_hook_fires_after_reboot() {
 
     # Restart the unit agent and ensure that the implicit start hook still
     # does not fire
+    wait_for_systemd_service_files_to_appear "ubuntu-lite/0"
```
Comment (achilleasa) [inline tests/suites/hooks/reboot.sh:29]: This was causing issues on (slow) CI as there was a small race between the unit appearing as idle and the systemd unit files being written to disk. This change addresses the issue and allows us to ensure that the start hook does not fire when the agent gets restarted _externally_.

File: tests/suites/hooks/reboot.sh [hunk 3]
```diff
@@ -39,6 +40,22 @@ run_start_hook_fires_after_reboot() {
       exit 1
     fi
 
+    # Ensure that the implicit start hook does not fire after upgrading the unit
+    juju upgrade-charm ubuntu-lite --revision 7
```
Comment (achilleasa) [inline tests/suites/hooks/reboot.sh:44]: This was backported from #12250

File: tests/suites/hooks/reboot.sh [hunk 4]
```diff
@@ -87,7 +104,8 @@ run_reboot_monitor_state_cleanup() {
     echo "[+] Verifying that reboot monitor state files are removed once a subordinate gets removed"
     juju remove-relation rsyslog-forwarder mysql
     wait_for "mysql" "$(idle_condition "mysql")"
-    sleep 5 # there is probably a better way to wait for the subordinate to be removed.
+
+    wait_for_subordinate_count "mysql"
```
Comment (achilleasa) [inline tests/suites/hooks/reboot.sh:108]: Also backported from #12250

**Review by petevg — COMMENTED** [review]
I am a little bit sad that we don't have cleaner methods for waiting for things. But definitely approve of hiding the messy bits in functions that can be made prettier later/when we have extended our abilities to wait for cloud states.

**Review by hmlanigan — APPROVED** [review]
LGTM, qa went well

**Comment by achilleasa** [comment]
Doh! This was supposed to be targeting 2.8....

**Pull request merged**